### ISSUE TITLE
Qwen3.8-27B-MTP-4bit drafter ships NaN weights → 0% speculative acceptance (BF16 drafter works, 82%)

### ISSUE THREAD

**Issue by L2ncE** [Blaizzy/mlx-vlm#1931]
## Summary

The published drafter checkpoint [`mlx-community/Qwen3.8-27B-MTP-4bit`](https://huggingface.co/mlx-community/Qwen3.8-27B-MTP-4bit) contains NaN/Inf values in its floating-point tensors (scales, biases, and RMSNorm weights). The poisoned drafter makes speculative decoding **always draft token 0** (argmax of NaN logits), producing a **0.0% acceptance rate** — MTP actually makes generation ~23% *slower* than plain AR.

The **BF16 version** ([`mlx-community/Qwen3.8-27B-MTP-bf16`](https://huggingface.co/mlx-community/Qwen3.8-27B-MTP-bf16)) is clean and works perfectly on the same target model and the same mlx-vlm code: **82.1% acceptance, 25.4 tok/s vs 14.6 tok/s AR baseline (1.74x)** on an M4 Pro.

## Environment

- Apple M4 Pro (24 GB unified memory), macOS 26.5.2
- mlx-vlm 0.6.13 (identical results on current `main` @ 2026-08-16 and on the PR #1899 branch)
- Target: `mlx-community/Qwen3.8-27B-4bit` (verified byte-size-identical to upstream)

## Symptom

```bash
mlx_vlm.generate --model mlx-community/Qwen3.8-27B-4bit \
  --draft-model mlx-community/Qwen3.8-27B-MTP-4bit \
  --prompt "Write a quicksort in Python." \
  --max-tokens 400 --enable-thinking --verbose
```

```
Generation: 400 tokens, 11.516 tokens-per-sec     (AR baseline: 14.6)
Speculative decoding: 1.00 accepted tokens/round (0.00 accepted drafts/round,
0.0% of drafted, avg draft 2.00) over 399 rounds
```

The result is identical with `--temperature 0`, `--draft-block-size 4`, and with/without `--enable-thinking`.

## Root cause

Direct inspection of the raw safetensors (bypassing the loader) shows 10 keys with NaN/Inf:

```python
import mlx.core as mx, numpy as np
w = mx.load("model.safetensors")   # the 4bit drafter
# → 10 floating-point keys contain NaN/Inf:
#   fc.biases                              nan=1651 inf=12
#   layers.0.input_layernorm.weight        nan=7
#   layers.0.mlp.gate_proj.scales          nan=2829 inf=14
#   layers.0.post_attention_layernorm.weight nan=10
#   layers.0.self_attn.k_proj.scales       nan=178
#   layers.0.self_attn.o_proj.biases       nan=968 inf=3
#   layers.0.self_attn.q_proj.scales       nan=88
#   layers.0.self_attn.q_proj.biases       nan=60
#   layers.0.self_attn.v_proj.biases       nan=149
#   layers.0.self_attn.v_proj.scales       nan=155
```

Instrumented drafter forward: inputs are healthy (hidden RMS ≈ 1.8), but the first `fc` output already contains ~31% NaN, which then propagates. Consequently `draft_block()` always returns `[0, 0]` for every input, and every draft is rejected.

Note: replacing only the NaN values in the floating-point tensors with medians is **not** sufficient — NaN still reappears in the dequantized output, suggesting some packed 4-bit groups are corrupted as well. The 4bit artifact needs a full re-conversion.

## BF16 control experiment (works)

| Drafter | Acceptance | Speed (decode tok/s) |
|---|---|---|
| `Qwen3.8-27B-MTP-4bit` | 0.0% | 11.5 (slower than AR) |
| `Qwen3.8-27B-MTP-bf16` | **82.1%** | **25.4 (1.74x over AR)** |

Same target, same code, same prompt, greedy (`--temperature 0`).

## Suggested fixes

1. Re-convert `Qwen3.8-27B-MTP-4bit` (the `mlx_vlm.convert` path appears to produce NaN in scales/biases for this MTP head — possibly a quantization overflow on near-zero scales).
2. Add a NaN/Inf sanity check on converted artifacts before upload (cheap: one pass over floating-point tensors).
3. Optionally, have `load_drafter` warn when the drafter output distribution is degenerate (constant argmax), which would make this failure mode obvious to users.

Happy to provide any additional diagnostics from this machine.

**Comment by UmarFarooq-Ch** [comment]
for me it is working, with same command you provided. Using M4 Pro 48GB

```
Prompt: 59 tokens, 83.372 tokens-per-sec
Generation: 400 tokens, 25.048 tokens-per-sec
Peak memory: 17.437 GB
Speculative decoding: 2.70 accepted tokens/round (1.70 accepted drafts/round, 85.4% of drafted, avg draft 1.99) over 148 rounds
```

**Comment by Lazarus-931** [comment]
hi @L2ncE,

I tried and it works just fine for me. 

maybe you have old `mlx-vlm` version?

**Comment by matthewalexong** [comment]
Additional M1 Max datapoint for the same prompt (`Write a quicksort in Python.`). I tested on a Mac Studio with M1 Max (64 GB), macOS 26.6.2, `mlx-vlm` 0.7.2 / `mlx` 0.32.2 / `transformers` 5.17.0. The target was `mlx-community/Qwen3.8-27B-4bit` at revision `10c35caafbb80f7dc6a7a432cdd11af10a6d4818`; the MTP drafter was `mlx-community/Qwen3.8-27B-MTP-4bit` at revision `b643c01b6d3b094e325edb6ebd832e16c486c575` (safetensors SHA-256 `76663c101e7e8ea9c0ae17bcb95183cd7f733ce424c912b8b264a7b1c48e4cc6`).

With greedy decoding (`temperature=0`), thinking enabled, and `max_tokens=400`, the model stopped at EOS after 287 tokens. Across three paired runs, MTP acceptance was 70.2%; median generation speed was 18.602 tok/s with MTP vs 19.171 tok/s autoregressively (about 3.0% slower). Peak memory was 17.114 GB vs 16.425 GB, and the visible greedy completion matched. I scanned the adapter's 23 floating-point tensors and found no NaN/Inf; this scan does not validate every packed 4-bit group.

This does not reproduce the reported M4 Pro / `mlx-vlm` 0.6.13 failure. It is a successful speculative-decoding run on a different Mac/runtime, but not a speedup on this M1 Max.
